cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java

```diff
@@ -169,18 +201,33 @@ public void setProperty(final MetadataScope scope, final MetadataEntity metadata
   @Override
   public void addTags(final MetadataScope scope, final MetadataEntity metadataEntity,
                       final Set<String> tagsToAdd) {
+    throwErrorOnV1Writes(scope, metadataEntity);
+
     final AtomicReference<MetadataRecordV2> previousRef = new AtomicReference<>();
     execute(mds -> {
       Map<String, String> existingProperties = mds.getProperties(metadataEntity);
       Set<String> existingTags = mds.getTags(metadataEntity);
       previousRef.set(new MetadataRecordV2(metadataEntity, scope, existingProperties, existingTags));
       mds.addTags(metadataEntity, tagsToAdd);
-    }, scope);
+    }, scope, DATASET.V2);
+
     publishAudit(previousRef.get(),
                  new MetadataRecordV2(metadataEntity, scope, EMPTY_PROPERTIES, Sets.newHashSet(tagsToAdd)),
                  new MetadataRecordV2(metadataEntity, scope));
   }
 
+  private void throwErrorOnV1Writes(MetadataScope scope, MetadataEntity metadataEntity) {
+    updateInMemoryState();
+
+    // we do not support writes to V1 tables during migration. So throw an error if we are trying to update
+    // an existing entity
+    if ((scope.equals(MetadataScope.SYSTEM) ? hasV1SystemDs : hasV1BusinessDs)
+      && hasEntityInV1(scope, metadataEntity)) {
+      throw new RuntimeException("Metadata migration is in progress. Please retry the same operation " +
```

Comment:
What will the HTTP code be? Is it a 500?

Reply:
yes it is 500 in case of `Exception`

Reply:
Why 500? It is more like operation not available, which should be `503 Service Unavailable`, meaning you should be throwing `ServiceUnavailableException` instead.